velox/exec/SortBuffer.cpp

```diff
@@ -0,0 +1,360 @@
+      spillMemoryThreshold_(spillMemoryThreshold) {
+  VELOX_CHECK_GT(input_->size(), sortCompareFlags_.size());
+  VELOX_CHECK_GT(sortCompareFlags_.size(), 0);
+  VELOX_CHECK_EQ(sortColumnIndices.size(), sortCompareFlags_.size());
+  VELOX_CHECK_GT(outputBatchSize_, 0);
+  VELOX_CHECK_NOT_NULL(nonReclaimableSection_);
+
+  std::vector<TypePtr> sortedColumnTypes;
+  std::vector<TypePtr> nonSortedColumnTypes;
+  std::vector<std::string> sortedSpillColumnNames;
+  std::vector<TypePtr> sortedSpillColumnTypes;
+  sortedColumnTypes.reserve(sortColumnIndices.size());
+  nonSortedColumnTypes.reserve(input->size() - sortColumnIndices.size());
+  sortedSpillColumnNames.reserve(input->size());
+  sortedSpillColumnTypes.reserve(input->size());
+  std::unordered_set<column_index_t> sortedChannelSet;
+  // Sorted key columns.
+  for (column_index_t i = 0; i < sortColumnIndices.size(); ++i) {
+    columnMap_.emplace_back(IdentityProjection(i, sortColumnIndices.at(i)));
+    sortedColumnTypes.emplace_back(input_->childAt(sortColumnIndices.at(i)));
+    sortedSpillColumnTypes.emplace_back(input_->childAt(sortColumnIndices.at(i)));
+    sortedSpillColumnNames.emplace_back(input->nameOf(sortColumnIndices.at(i)));
+    sortedChannelSet.emplace(sortColumnIndices.at(i));
+  }
+  // Non-sorted key columns.
+  for (column_index_t i = 0, nonSortedIndex = sortCompareFlags_.size();
+       i < input_->size();
+       ++i) {
+    if (sortedChannelSet.count(i) != 0) {
+      continue;
+    }
+    columnMap_.emplace_back(nonSortedIndex++, i);
+    nonSortedColumnTypes.emplace_back(input_->childAt(i));
+    sortedSpillColumnTypes.emplace_back(input_->childAt(i));
+    sortedSpillColumnNames.emplace_back(input->nameOf(i));
+  }
+
+  data_ = std::make_shared<RowContainer>(
+      sortedColumnTypes, nonSortedColumnTypes, pool_);
+  spillerStoreType_ =
+      ROW(std::move(sortedSpillColumnNames), std::move(sortedSpillColumnTypes));
+}
+
+void SortBuffer::addInput(const RowVectorPtr& input) {
+  VELOX_CHECK(!noMoreInput_);
+  ensureInputFits(input);
+
+  // Prevents the memory arbitrator to reclaim memory from this SortBuffer
+  // during the execution below.
+  //
+  // NOTE: 'nonReclaimableSection_' points to the corresponding flag in the
+  // associated OrderBy/TableWriter operator.
+  auto guard = folly::makeGuard([this]() { *nonReclaimableSection_ = false; });
+  *nonReclaimableSection_ = true;
+
+  SelectivityVector allRows(input->size());
+  std::vector<char*> rows(input->size());
+  for (int row = 0; row < input->size(); ++row) {
+    rows[row] = data_->newRow();
+  }
+  for (const auto& columnProjection : columnMap_) {
+    DecodedVector decoded(
+        *input->childAt(columnProjection.outputChannel), allRows);
+    for (int i = 0; i < input->size(); ++i) {
+      data_->store(decoded, i, rows[i], columnProjection.inputChannel);
+    }
+  }
+  numInputRows_ += allRows.size();
+}
+
+std::optional<SpillStats> SortBuffer::noMoreInput() {
+  VELOX_CHECK(!noMoreInput_);
```

Comment:
I assume there is no logic change in these function? thanks!

Reply:
Correct